### ISSUE TITLE
experiment: run the matched PPO versus RecurrentPPO recurrence ablation

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#7849]
## Current issue-planning review — 2026-09-15

**Audit disposition:** `blocked`. Active open PR evidence is present; the issue remains open and no ownership change is made.

**First action:** Resolve #7846 preregistration, #7847 runtime and #7848 stateful adapter exact-head receipts before any job allocation.

### Ordered implementation / investigation plan

1. Resolve #7846 preregistration, #7847 runtime and #7848 stateful adapter exact-head receipts before any job allocation.
2. Freeze two arms, five training seeds and the declared 15M steps per run only if the final contract retains that budget; check observation/action/split/checkpoint-selection parity.
3. After explicit compute/domain authorization, run a disjoint infrastructure canary then the frozen ten training runs; monitor checkpoint/progress/state integrity without tuning from held-out outcomes.
4. Evaluate selected checkpoints on the complete paired held-out matrix, preserve training-seed clustering and missingness, and record continue/revise/stop/inconclusive.

### Files and owners to inspect

- configs/training/comparison_matrix/
- configs/training/ppo/
- scripts/training/
- SLURM/ and private-ops under their existing authority

### Acceptance and validation

- No wrapper, reward change, replacement failed seeds or test-set checkpoint selection.
- All checkpoints and terminal outcomes have durable checksums; actions and recurrent resets agree with deployment.

Named validation commands:

- `scripts/training/`
- `scripts/analysis/`

### Current dependencies and issue-text corrections

- Dependencies: #7846 #7847 #7848; explicit compute/domain authorization
- Keep blocked/compute-needed rather than implementation-ready until exact admission exists.
- Record current status of each predecessor and freeze actual commands instead of leaving executable placeholders.
- The supplied draft was checked against the refreshed live issue and complete comment thread; historical rationale is retained below.
- Keep current out-of-scope, ownership, provenance, rights, compute, and claim boundaries intact. Do not close the issue merely because this plan was posted.

### Lifecycle and claim boundary

- Current state labels: `state:running`.
- The issue remains open. No closure, compute submission, release/publication, evidence deletion, or scientific-claim promotion is authorized by this review.
- This review does not request a lifecycle transition; existing assignees, ownership, valid priority labels, and unrelated labels are preserved.
- Proposed non-lifecycle label additions applied with this review: `dependency:has-blockers`.

**Evidence note:** Full live issue body and complete available comment thread were reviewed during the 2026-09-15 refresh. This is planning metadata, not implementation or benchmark evidence; historical measurements remain historical.

## Parent and Dependencies

- Programme: #7845
- Preregistration/answerability: #7846
- Training runtime: #7847
- Stateful planner adapter: #7848
- Position: child D and the first evidence-producing campaign.

## Archetype Metadata

```yaml
archetype: training-and-evaluation-campaign
evidence_tier: full-matrix-candidate
resource: slurm-gpu
execution_authorized: false
```

## Objective

Execute the frozen two-arm, five-training-seed PPO versus RecurrentPPO comparison from #7846 after all implementation, answerability, exact-source, artifact, and Domain-Aware Approval gates pass.

This campaign tests recurrence only. It must not enable the safety wrapper, alter observations, change rewards, tune hyperparameters from outcomes, replace failed seeds, or add another policy architecture.

## Research Question

Under the exact matched contract, does true RecurrentPPO improve collision-free route completion or reduce low-progress/timeout failure relative to feed-forward PPO without a material collision or severe-intrusion regression?

## Frozen Arms

The exact identifiers come from #7846. Conceptually:

1. feed-forward PPO control on `default_gym` `drive_state + rays`;
2. true `sb3_contrib.RecurrentPPO` with `MultiInputLstmPolicy` on the same observation and action contract.

Both arms must share:

- training and evaluation scenario manifests;
- reward and termination semantics;
- desired `(v, omega)` command contract;
- canonical `dt`-correct conversion to acceleration;
- robot dynamics and limits;
- training steps per seed;
- training-seed list;
- validation/evaluation cadence;
- selected-checkpoint rule;
- held-out evaluation identities;
- artifact and failure policy.

Only the recurrence-specific algorithm/policy state may differ, apart from unavoidable parameter-count and computational differences that are reported rather than hidden.

### Gate 0: exact-source admission

Before any job submission:

- freeze one immutable public source SHA;
- prove exact config, scenario, split, observation, action, dependency, and artifact-destination digests;
- run #7846's answerability and parity validators;
- run #7847's bounded training/evaluation smoke;
- run #7848's state-lifecycle, deterministic replay, and action-parity smoke;
- prove both arms construct natively with no fallback/degraded mode;
- record explicit maintainer Domain-Aware Approval and compute authorization.

### Gate 1: non-evidence infrastructure canary

Use disjoint non-evidence identities frozen in the launch packet to prove:

- both arm commands start and make environment-step progress;
- recurrent state is finite and resets correctly;
- periodic evaluation and checkpoint writes work;
- scheduler route, GPU/CPU memory, artifact upload, and terminal finalization work;
- no action-space, observation, checkpoint, or source mismatch exists.

The canary is implementation evidence only. Do not use its policy outcomes for hyperparameter, scenario, threshold, seed, budget, or checkpoint-rule selection.

### Gate 2: full matched training

Unless #7846 freezes a different outcome-free justified budget, execute:

```text
2 policy arms x 5 independent training seeds x 15,000,000 environment steps
= 10 independently attributable training runs
= 150,000,000 total planned environment steps
```

Training seeds are `123, 231, 777, 992, 1337` unless the final #7846 contract records a reviewed successor before outcomes.

Every arm/seed run must preserve all scheduled checkpoints and evaluation summaries required by the frozen rule. An infrastructure retry requires a changed environment/packet and an explicit classification. Scientific failures and unfavorable outcomes are not retryable.

### Gate 3: held-out evaluation

For every selected checkpoint:

- evaluate the complete frozen evaluation matrix;
- use identical scenario/seed identities for paired arm comparisons;
- keep training-seed identity as an independent replication level;
- use deterministic inference;
- reset recurrent state at every episode/scenario boundary;
- preserve raw commands, converted actions, outcomes, and state-integrity diagnostics;
- reject fallback, degraded, unavailable, failed, duplicate, missing, and provenance-invalid rows from success evidence while retaining them in the ledger.

No final-test row may be used for checkpoint selection or training decisions.

## Primary and Secondary Analysis

The exact minimally important effects, collision margin, uncertainty method, and decision predicates are those frozen in #7846.

Required analysis structure:

1. Account for all planned training runs and evaluation cells before aggregation.
2. Report each training seed separately.
3. Estimate paired arm differences within matched scenario/evaluation-seed cells.
4. Preserve training seed, scenario, and evaluation seed as distinct clustering levels.
5. Use the preregistered hierarchical paired bootstrap or approved equivalent.
6. Report episode and scenario counts, missingness, attrition, native/fallback/degraded status, and uncertainty.
7. Report overall results and scenario-family/failure-mechanism strata.
8. Keep collision-free completion primary and collision/severe-intrusion constraints separate.
9. Treat time-to-goal and efficiency as conditional or secondary where required by the contract.
10. Report computational cost and latency independently from navigation quality.

Required outputs include:

- collision-free route completion;
- success and collision probabilities;
- typed collisions;
- near-miss/exposure and minimum-clearance outcomes;
- low-progress, timeout, and deadlock rates;
- time to goal conditional on collision-free completion;
- path efficiency;
- acceleration, jerk, command saturation, and realized-speed summaries;
- median/p95/max inference latency;
- training throughput, wall time, memory, and checkpoint sizes;
- recurrent reset/state-integrity diagnostics;
- parameter counts.

## Live Monitoring and Watcher Contract

Every submitted job must have an adaptive watcher and a terminal finalizer.

### Launch and provenance checks

At launch, record and verify:

- repository and exact source SHA;
- clean/dirty state;
- resolved config and all input digests;
- training arm and seed;
- scheduler partition, GPU type/count, CPU, RAM, and wall-time request;
- package and driver versions;
- output root and durable artifact URI;
- job ID and launch command digest.

### Progress monitoring

Track at bounded intervals:

- scheduler state and allocation host;
- environment steps and update count;
- wall time and estimated progress fraction;
- steps/s and evaluation duration;
- latest checkpoint step, age, checksum status, and size;
- CPU RAM and GPU allocated/reserved/peak memory;
- worker liveness and subprocess exits;
- optimization metrics and non-finite counts;
- action saturation and zero-motion diagnostics;
- recurrent state finite/reset/leakage diagnostics;
- periodic evaluation success, collision, low-progress/timeout, near miss, and latency.

Polling cadence may widen after stable progress, but terminal state ingestion must be exact. Monitoring must not inspect final held-out outcomes early for design changes.

### Alert and stop conditions

Stop or hold the affected run on:

- no environment-step or checkpoint progress beyond the launch-packet stall threshold;
- non-finite losses, parameters, actions, observations, or recurrent state;
- unexpected recurrent state carry-over or wrong environment ownership;
- persistent zero-motion or constant action saturation under the frozen diagnostic rule;
- observation/action/checkpoint/source drift;
- OOM, repeated worker crash, corrupted checkpoint, storage failure, or artifact-upload failure;
- evaluation identity, cardinality, or state-reset failure;
- duplicated or overlapping run identity;
- unavailable durable destination.

Do not automatically launch a replacement. Classify the failure and require an updated differential packet.

## Artifact and Custody Contract

Large outputs remain in approved durable storage. Each run must publish or reference:

- resolved config and launch packet;
- source/dependency/environment manifest;
- training/evaluation metrics;
- checkpoint index and checksums;
- selected checkpoint and rule proof;
- terminal scheduler/process receipt;
- recurrent-state integrity summary;
- raw held-out episode rows or immutable durable pointer;
- complete row/missingness ledger;
- compact analysis packet;
- `SHA256SUMS` or equivalent digest inventory.

The public repository may track compact manifests, analysis outputs, evidence-review markers, and durable pointers. Do not commit raw checkpoints or large traces.

## Result Classification

Apply exactly one final result from #7846:

- `continue_to_wrapper_factorial`;
- `revise_recurrent_contract`;
- `stop_recurrent_direction`;
- `inconclusive_invalid_or_incomplete`.

Examples:

- A success gain with collision non-inferiority under the frozen rule may continue.
- A collision regression cannot be compensated by success, comfort, efficiency, or SNQI.
- Lower timeout without a primary-result improvement remains a secondary finding.
- A recurrent-state defect, incomplete matrix, or provenance failure is invalid/inconclusive, not a negative policy result.
- A valid null or negative result is terminal evidence and must be reported.

## Implementation and Execution Guidelines

- Use only merged canonical runners and adapters from #7847/#7848.
- Do not patch runtime code after the first registered outcome. A defect requires stopping and versioning a new campaign.
- Keep one immutable campaign source SHA for every arm and seed.
- Keep arm and seed outputs separate until analysis.
- Do not perform outcome-dependent early stopping unless #7846 explicitly preregisters one identical rule for both arms.
- Do not tune reward, architecture, LSTM size, rollout length, scenario weights, or checkpoint metric inside this campaign.
- Do not enable the safety wrapper, ORCA residual, CBF, predictive foresight, or privileged observations.
- Do not count failed, fallback, degraded, unavailable, or missing rows as zero-valued evidence.

## Suggested Paths

Preparation and compact result work may use:

- `configs/training/comparison_matrix/`
- `configs/training/ppo/`
- `scripts/training/`
- `scripts/analysis/` or the canonical analysis owner
- `SLURM/` and private-ops launch surfaces under their existing authority
- `docs/context/evidence/`
- focused campaign/analysis tests

Do not change planner, metric, reward, scenario, or safety-wrapper implementations from this execution issue.

## Acceptance Criteria

- [ ] #7846 is merged, independently reviewed as answerable, and exact bytes/digests are frozen.
- [ ] #7847 and #7848 are merged with passing smoke and exact action/state contracts.
- [ ] Domain-Aware Approval and compute authorization name this exact campaign.
- [ ] Gate-1 canary passes for both arms without fallback/degraded execution.
- [ ] All ten planned training runs execute or receive explicit fail-closed terminal classifications.
- [ ] Every selected checkpoint follows the frozen selection rule and has durable checksum-covered lineage.
- [ ] The complete held-out evaluation matrix is accounted for before aggregation.
- [ ] Recurrent state resets and ownership are valid for every recurrent evaluation row.
- [ ] Analysis follows the frozen cluster, uncertainty, margin, and decision contract.
- [ ] Raw and compact evidence have durable custody and verified digests.
- [ ] The final disposition is posted to #7845 with exact artifacts and limitations.
- [ ] No safety-wrapper, default-planner, release, dissertation, or paper claim is changed automatically.

## Validation

The exact launch and analysis commands must be generated by #7846/#7847 and copied into the final launch packet. Before submission, run at least:

```bash
<contract validator --check>
<research answerability check>
<feed-forward dry-run and smoke>
<recurrent dry-run and smoke>
<stateful adapter deterministic replay smoke>
<campaign manifest cardinality and digest check>
```

After completion:

```bash
<row-accounting validator>
<registered paired analysis command>
<artifact checksum verification>
<independent exact-digest evidence review>
```

## Stop Conditions

- Stop before submission if any dependency, source, model, observation, action, seed, split, artifact, or answerability predicate is unresolved.
- Stop after any runtime defect that would require code/config changes.
- Stop if the matrix cannot finish without outcome-dependent replacement.
- Stop if held-out results are inspected before the frozen stopping point.
- Do not continue to wrapper testing unless the exact final disposition permits it.

## Claim Boundary

A valid result supports only a bounded comparison of the named feed-forward and recurrent implementations under the frozen Robot SF simulator, observation, action, training, scenario, seed, and evaluation contracts. It does not establish universal recurrence superiority, real-world prediction, deployment safety, or sim-to-real transfer.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 7849
source_body_sha256: 5e0bd4a3eb7956d5f2495ced3e1c2e24cd97503fd1115144dc34a9116558b790
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 1c191411793929640684e0827d763f3a2009a8682ab98dcefa8c2ded3e54f4ea
audit_classification: needs_compute
next_action: route_to_compute_authority
authority: compute_authority
execution_mode: compute
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-20260830-04
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.
